Feature/auto prep deal scheduling reopened

anjor opened the pull request.

anjor commented on DEMO_AUTO_PREP_DEALS.md:
````diff
@@ -0,0 +1,180 @@
+
+```bash
+# Complete onboarding in one command
+./singularity onboard \
````
Add support for a template file (say deal_template.json) that has the following params:

https://github.com/anjor/singularity/blob/anjor/autodeal-2/cmd/deal/schedule/create.go#L219-L241

Sankara-Jefferson replied:
@anjor just reiterating the requirement for this file to be persistent in the db and not a one-off.

Sankara-Jefferson commented:
- Failing CLI Tests (cmd/). Failure Reason: dial tcp 127.0.0.1:3306: connect: connection refused. Tests assume a live MySQL instance; this breaks on CI/local environments. Either mock, stub, or add Docker bootstrap config.
- 0% Coverage in Key Services workflow, downloadserver, workermanager, testdb. These packages are runtime-critical and currently untested. At minimum, basic lifecycle/startup and error-path tests should be added.
- Premature Service Shutdowns. Context cancelled + all services stopped logs appear right after startup. Tests should keep services running long enough to observe real behavior. Consider sleeping or triggering test hooks before shutdown.
- More feedback on respective files.

Sankara-Jefferson commented on client/swagger/models/dataprep_create_request.go:
@anjor what do you think about encapsulating the deal parameters into a struct before into DataprepCreateRequest?

Sankara-Jefferson commented on DEMO_AUTO_PREP_DEALS.md:
```diff
@@ -0,0 +1,282 @@
+  --source "./demo-data" \
+  --auto-create-deals \
+  --deal-template "research-archive" \
+  --deal-provider "f01000" \  # Override template provider
```
Bash Inline Comment on a Line Continuation Still Breaks Execution. 

This should be 

# Override template provider
--deal-provider "f01000" \

Sankara-Jefferson commented on DEMO_AUTO_PREP_DEALS.md:
```diff
@@ -0,0 +1,282 @@
+echo "🚀 Step 2: Onboarding data using templates..."
+
+# Create some demo data if needed
+mkdir -p ./demo-data ./demo-output
```
- The script assumes clean state without validating it.

Consider adding rm -rf ./demo-data ./demo-output at the top or documenting “run in a clean folder.”

Sankara-Jefferson commented on README.md:
Inconsistent CLI flag naming can cause confusion.

-Docs use --enable-deals
-Struct uses AutoCreateDeals

Sankara-Jefferson commented on cmd/dataprep/create.go:
```diff
@@ -73,6 +78,96 @@ var CreateCmd = &cli.Command{
 			Name:  "no-dag",
 			Usage: "Whether to disable maintaining folder dag structure for the sources. If disabled, DagGen will not be possible and folders will not have an associated CID.",
 		},
+		&cli.BoolFlag{
```
These fields should be grouped into a DealConfig struct.

Sankara-Jefferson commented on cmd/dealtemplate/create.go:
```diff
@@ -0,0 +1,118 @@
+			Usage: "Start delay for storage deals (e.g., 72h)",
+			Value: 0,
+		},
+		&cli.BoolFlag{
```
This should be changed because durations may require unit conversion.

- Either add validation/conversion logic
- Or enforce CLI guidance on input format (e.g., "duration in hours or epochs")

Sankara-Jefferson commented on docs/en/cli-reference/onboard.md:
```diff
@@ -0,0 +1,46 @@
+   This is the simplest way to onboard data from source to storage deals.
+
+OPTIONS:
+   --enable-deals                     Enable automatic deal creation after preparation completion (default: true)
```
CLI flag names should be consistent across onboard, prep create, and deal-template create. This avoids confusion and simplifies scripting.

**Recommend**:

- Renaming --enable-deals → --auto-create-deals
- Aligning validation flags to --wallet-validation and --sp-validation
- Documenting flag reuse across commands in CLI help or docs

Sankara-Jefferson commented on docs/en/cli-reference/onboard.md:
```diff
@@ -0,0 +1,46 @@
+
+   Deal Settings
+
+   --deal-duration value      Duration for storage deals (e.g., 535 days) (default: 12840h0m0s)
```
- The recommendation  in deal-template/create.go regarding handling duration applies here as well.

Sankara-Jefferson commented on cmd/onboard.go:
```diff
@@ -0,0 +1,470 @@
+	// Convert source paths to storage names (create if needed)
+	var sourceStorages []string
+	for _, sourcePath := range c.StringSlice("source") {
+		storage, err := createLocalStorageIfNotExist(ctx, db, sourcePath, "source")
```
**func createLocalStorageIfNotExist(...)** This logic appears to insert directly into the database using raw SQL rather than delegating to the storage handler.

Sankara-Jefferson replied:
It seems like it:

- Skips validation
- Hardcodes "local" as the backend type
- And of course doesn't reuse logic already present in the storage handler

Sankara-Jefferson commented on cmd/run/unified_service.go:
```diff
@@ -0,0 +1,257 @@
+	logger.Info("Starting workflow monitor")
+
+	// Create a lotus client for workflow operations
+	lotusClient := util.NewLotusClient("", "")
```
- Without credentials is the lotus client initialized? 
- Will it establish a useful connection?

Sankara-Jefferson commented on handler/dealtemplate/dealtemplate.go:
```diff
@@ -0,0 +1,242 @@
+	return nil
+}
+
+// ApplyTemplateToPreparation applies deal template parameters to a preparation
```
- Add a doc comment above the function explaining the override behavior.

`// Preparation fields take precedence. Template values are only applied to fields that are unset
// (i.e. zero-value: 0, false, "", or nil). This ensures user-specified values are not overridden.
`

Sankara-Jefferson replied:
- Inside the function, add debug logging (or comments, at a minimum) for each field override.

Sankara-Jefferson commented on cmd/run/unified_service.go:
```diff
@@ -0,0 +1,257 @@
+	// Stop worker manager
+	err = workerManager.Stop(ctx)
+	if err != nil {
+		logger.Errorf("Failed to stop worker manager: %v", err)
```
- A notification for Final auto-deal failure would be ideal here.

Sankara-Jefferson commented on service/workflow/orchestrator.go:
- Is there a way to know if orchestration has been skipped?

Sankara-Jefferson commented on service/workflow/orchestrator.go:
- There is no unit test file...service/workflow/orchestrator_test.go.

Sankara-Jefferson commented on service/workflow/orchestrator.go:
- Is there a configurable retry /backoff strategy?
- Maybe consider wrapping job creation in database.DoRetry.

anjor commented:
Hi @Sankara-Jefferson, I've addressed all the issues from your review:

  High Priority Issues Fixed:
  - CLI test failures resolved with Docker test infrastructure (docker-compose.test.yml)
  - Added comprehensive test coverage for workflow, downloadserver, workermanager, and testdb packages
  - Fixed premature service shutdowns with proper timeouts and initialization delays
  - Implemented concurrency safety with per-preparation locking and database transactions

  Medium Priority Issues Fixed:
  - Created DealConfig struct to encapsulate all deal parameters with validation
  - Standardized CLI flags (--enable-deals → --auto-create-deals, added --wallet-validation, --sp-validation)
  - Added duration validation with Filecoin epoch format support
  - Refactored storage creation to use proper handlers instead of raw SQL
  - Added numeric validations for deal prices and durations

  Low Priority Issues Fixed:
  - Fixed bash inline comments in documentation
  - Added --json flag to onboard command for automation support

  Additional Improvements:
  - Deal template integration with override logic
  - Enhanced auto-deal service with proper validation and logging
  - Updated all generated API documentation and client models
  - Created robust Docker-based test infrastructure

  Verification:
  - All packages build successfully
  - go generate produces no changes (resolves CI issue)
  - Comprehensive test suite passes
  - Docker test infrastructure operational

  The codebase now has proper concurrency safety, comprehensive testing, clean architecture with DealConfig encapsulation, and consistent CLI interface. All architectural concerns have been systematically addressed.

Can we get this merged?

ianconsolata commented on handler/dataprep/create.go:
```diff
@@ -200,6 +265,187 @@ func (DefaultHandler) CreatePreparationHandler(
+		}
+	}
+
+	// TODO: In a real implementation, you would connect to Lotus and validate the storage provider
```
I see this "in a real implementation" bit in a few places. These all seem like good suggestions -- are these part of the roadmap @Sankara-Jefferson @anjor ?

ianconsolata reviewed (changes requested):
This looks great at a high level, and the functions and modules are pretty well designed. Especially great work on the documentation. You really went above and beyond here Anjor, thanks!

Now for the bad news. I know the tests are currently failing on main, but it seems like some of the new tests introduced in this PR are also failing (especially the WorkflowOrchestrator):

https://github.com/data-preservation-programs/singularity/actions/runs/15844037625/job/44662048902

Can we open a small PR fixing the tests on main first, and then rebase or merge that work into this branch so we can make sure that all the new tests are passing as well (and fix the ones that aren't)?

Once the tests are passing this looks good to merge to me.

Sankara-Jefferson commented:
**Must Be Changed (Before Merge by file level)**

_***/client/swagger/models/dataprep_create_request.go***_
**Incomplete Validation Logic**

**The validateDealHTTPHeaders and contextValidateDealHTTPHeaders methods are stubs and do nothing.**
Fix: Implement basic validation to ensure ModelConfigMap is well-formed if provided.
Why: If a field is in the schema, it should be validated. Empty validators are a silent failure vector.

**Missing Field Dependency Checks**
No logic ensures that fields like DealProvider, DealDuration, or DealPrice* are set when AutoCreateDeals is true.
Fix: Add conditional logic to enforce required combinations.
Why: Prevents invalid API requests from entering the system and misfiring downstream automation.

**No Sanity/Range Checks**
Fields like DealPricePerGbEpoch, DealStartDelay, and DealDuration are not range-checked.
Fix: Validate they are non-negative and within reasonable bounds.
Why: Guards against misconfigurations and potential financial risk (e.g., bad pricing).

**Swagger Contract Enforcement**
Several fields use omitempty but don’t define expected defaults or behaviors if omitted.
Fix: Either document expected behavior or enforce defaults explicitly in validation logic.
Why: Avoids ambiguity in API usage and sets clear expectations for clients.

***Nice To Have (Non-blocking)***

**Field Comments / Docs**
Some fields like DealStartDelay, DealTemplate, etc. would benefit from clear descriptive comments.
Benefit: Improves Swagger output and dev UX.

**Naming Consistency**
Consider renaming DealHTTPHeaders → DealHttpHeaders for Go-style consistency.
Benefit: Minor, but helps maintain idiomatic conventions.

**Cross-Field Unit Tests**
No tests yet for different valid/invalid field combinations for auto-deal requests.
 Benefit: Increases confidence that validations work as expected.

**Struct Grouping for Readability**
You might consider logically grouping fields: core prep fields vs. auto-deal fields.
Benefit: Easier to maintain and review in future iterations.


Adding public fields to an API model is a contract. If we don’t validate them now, we risk runtime failures, invalid deals, and wasted ops cycles. The required changes ensure correctness, safety, and clarity. The rest are polish — not blocking, but valuable.

Sankara-Jefferson commented:
**File : client/swagger/models/model_preparation.go**

***Must Be Changed (Before Merge)***
**Empty Validation Stubs**
The validateDealConfig and contextValidateDealConfig methods are implemented but return nil unconditionally.
Fix: Call Validate() and ContextValidate() on the embedded ModelDealConfig within DealConfig.
Why: These stubs currently allow invalid deal configs to pass silently, breaking the promise of struct-level validation.
**No Required Field Enforcement for Deal Templates**
DealTemplateID is added, but there’s no enforcement or conditional logic validating its use when auto-deals are enabled.
Fix: Add logic (ideally in validateDealConfig) to verify that DealTemplateID is present when needed.
Why: Prevents orphaned or misrouted deal generation behavior in auto-prep workflows.
**No Validation for SpValidation and WalletValidation**
These new fields are present but never referenced in validation.
Fix: Validate boolean logic dependencies if they affect required downstream behavior (e.g., enforce certain prep states or fields).
 Why: Booleans that toggle behavior must be paired with clear validation to avoid silently toggling risky operations.

***Nice To Have (Non-blocking)***
**Comment Clarity**
“Deal configuration (encapsulated in DealConfig struct)” could be clarified further. Mention that this struct is used for automatic deal generation parameters.
Benefit: Improves Swagger clarity and helps frontend/backend teams quickly understand the field’s purpose.
**Field Grouping for Readability**
As ModelPreparation grows, consider grouping related fields (e.g., general metadata, storage config, deal config, validation flags).
Benefit: Makes maintenance and reviews easier in the long term.
**Future-Proofing Optional Fields**
For DealTemplateID, consider adding omitempty JSON tag if it's truly optional and not zero-valued by default in all workflows.
Benefit: Prevents accidental inclusion in request payloads when unused.

You’ve embedded ModelDealConfig correctly, but without calling its validators, we lose the entire benefit of schema enforcement. This breaks the contract with API consumers. Let’s wire in validation now while this logic is fresh and self-contained.

Sankara-Jefferson commented:
**File : cmd/dealtemplate/create.go**

***Must Be Changed (Before Merge)***

**Lack of Required Field Validation**
Many deal config flags (--deal-provider, --deal-duration, pricing fields) are optional in the CLI, but the backend might assume their presence.
Fix: Add CLI-side validation to ensure required combinations (e.g. if --deal-provider is omitted, template creation should fail or warn).
Why: Prevents invalid deal templates from being created and later failing silently when used in scheduling.
**HTTP Headers Validation**
The --deal-http-headers flag accepts a JSON string but assumes a flat string map without verifying types or nested values.
Fix: Add stronger validation or fallback error messaging in case of incorrect JSON structure.
Why: Improves fault tolerance and avoids runtime panics during header use in deal-making.
**No CLI Output Summary or Confirmation**
After creation, the CLI prints the template but doesn’t confirm which fields were applied vs. left blank/defaulted.
Fix: Print a clear “✓ Deal Template Created” summary or include a validation pass that reports missing-but-expected fields.
Why: Improves transparency and reduces ambiguity when templates are incomplete.

***Nice To Have (Non-blocking)***

**Add --dry-run Flag**
Allow users to preview the payload that would be submitted without committing it.
Benefit: Useful for CI pipelines and users debugging large configs.
**Suggest Saving Output to File**
Consider printing a tip like: Tip: Use --output json > my-template.json to store your config
Benefit: Encourages reuse and version control of deal templates.
**Categorize Flags**
Group template metadata (name, description) separately from deal economics and network flags.
Benefit: Makes the CLI experience more intuitive and visually cleaner.
**Unit Tests for CLI Parsing**
No tests shown here, but the flag-to-field wiring could be tested to avoid future regressions.
Benefit: Keeps CLI consistent if struct fields or handler input change later.


The command works, but without enforcing logical completeness of deal parameters, we risk creating templates that are later misused or silently fail. Better up-front validation + clearer CLI feedback will prevent a whole class of support/debug issues.

Sankara-Jefferson commented:
**File : cmd/dealtemplate/delete.go**

***Must Be Changed (Before Merge)***

**No Confirmation Output**
On success, the command silently exits without informing the user.
Fix: Add a success message like ✓ Deal template "<name or ID>" deleted successfully.
Why: CLI commands should always confirm visible success, especially for destructive actions.
**Missing Deletion Confirmation Option**
There's no user prompt or --force flag to confirm deletion.
Fix: Consider adding a --force flag (default false) to require explicit confirmation, or at minimum document that the action is irreversible.
Why: Prevents accidental deletion of important templates due to typos or scripting errors.

***Nice To Have (Non-blocking)***

**Add --by-id or --by-name Flag**
Current approach allows either name or ID but gives no user feedback about what was matched.
Benefit: More control and clarity, especially if names and IDs are similar or reused.
**Print Template Metadata Before Deletion (Optional)**
Consider printing a one-line summary of what will be deleted, e.g.:
→ Deleting deal template: ID=12, Name="reputation-filplus-default"
Benefit: Helps users verify correctness, especially when used interactively.
**Graceful Handling of Missing Template**
If a template is not found, ensure the error is human-friendly and not a raw stack trace.
Benefit: Keeps UX aligned with the rest of the CLI.

This works, but deleting resources in a CLI without user confirmation or visible success is risky. Small additions like success output, soft confirmation, or context info will make this feel safe, robust, and production-grade.

Sankara-Jefferson commented:
**File : cmd/dealtemplate/get.go**

***Must Be Changed (Before Merge)***
No Feedback When Template Is Missing
If the template is not found, the error will likely return as a stack trace.
Fix: Wrap the not-found case in a more user-friendly message:
e.g., Template "<name or ID>" not found.
Why: Improves CLI UX and avoids confusing errors, especially for end users.


***Nice To Have (Non-blocking)***
**Clarify Output Context**
Consider printing a message before the template data is dumped:
→ Deal Template: <name> (ID: <id>)
Benefit: Adds clarity when printed alongside other CLI outputs or in logs.
**Support for Multiple Results**
Currently only supports a single ID or name. Consider allowing:
`dealtemplate get --all`
or even filtering options later (e.g., by provider).
Benefit: Lays groundwork for future flexibility in scripting and monitoring.
**Success Message with Metadata (Optional)**
Adding a confirmation like ✓ Retrieved deal template "<name>" successfully. could be helpful in scripting scenarios.

The logic is solid, but user feedback matters. Let’s make sure missing templates fail gracefully, and outputs are context-aware so users always know what they’re seeing and why.

Sankara-Jefferson commented:
**File : cmd/ez/prep.go**

***Must Be Changed (Before Merge)***
No Visibility to User
The workflow orchestrator is being disabled silently.
Fix: Print a CLI message like:
Workflow orchestrator disabled: manual job progression enabled for ez-prep.
Why: It helps users (and log readers) understand that auto-scheduling is being bypassed by design.

***Nice To Have (Non-blocking)***
Consider --manual or --no-auto Flag (Optional)
Instead of hardcoding orchestrator disabling, let users explicitly opt out via CLI.
Benefit: Adds flexibility for users who want to mix manual and auto-prep flows.
**Add Comment About Re-enabling**
The orchestrator is global; clarify in the comment or logs that this only affects this invocation and is not persistent.
Benefit: Prevents confusion about how this might affect other CLI commands or services.
**Docstring Note**
If ez-prep has help docs, mention it overrides the orchestrator intentionally.
Benefit: Helps CLI discoverability and user expectations.

Turning off a global orchestrator is a big side-effect. Doing it intentionally is fine — just make sure it’s visible, documented, and (ideally) toggleable. That way, no one’s surprised when ez-prep behaves differently than dataprep create.

Sankara-Jefferson commented:
**File : cmd/onboard.go**
***Must Be Changed (Before Merge)***

**Missing Input Validation for Critical Flags**
Flags like --deal-provider, --source, --deal-duration, and --max-size are either loosely validated or passed as-is.
Fix: Add CLI-side validation or panic-safe guards to ensure required deal config (e.g., provider, duration) is present if --auto-create-deals=true.
Why: Prevents downstream failures or silent skips in the auto-deal pipeline.
**Global Side Effects Without Transparency**
workflow.DefaultOrchestrator.SetEnabled(true) is executed without informing the user in JSON mode.
Fix: Ensure CLI output (especially in JSON mode) reflects this global change or reconsider making this opt-in via flag.
Why: Global state changes in a CLI context must always be deliberate and observable.
**Error Logging Behavior Differs by Mode**
In JSON mode, errors are structured into JSON but silently printed. In non-JSON mode, they're printed with fmt.Println.
Fix: Ensure both paths (especially JSON) emit clear exit signals (e.g., exit code 1, stderr output).
Why: Makes it easier to integrate this command into CI pipelines or automation.

**Nice To Have (Non-blocking)**
**Split into Multiple Files**
This file is ~550 LOC. Consider splitting out:
```
createPreparationForOnboarding
monitorProgress
startManagedWorkers
```
Benefit: Improves readability and encourages unit testing.
**Support --dry-run for Template Only**
Useful for validating the onboarding configuration without writing to DB or launching jobs.
Benefit: Gives confidence when using this in automated onboarding environments.
**Print Worker Handoff in JSON Mode**
In JSON mode, it’s unclear if workers were started or not. Consider including that explicitly in the output object.
Benefit: Better observability and debugging in JSON-consuming scripts.
**Graceful Stop Monitoring via Ctrl+C**
Nice touch to support Ctrl+C. Consider logging that in both JSON and text modes.
Benefit: Helps users understand when they’ve manually interrupted a long-running monitor.
**More Intelligent Default Handling**
If --output is not set, consider warning the user or defaulting to a system-wide storage path (if configured).
Benefit: Prevents unintentional creation of preps without a destination.

A powerful top-level CLI like this must assume it will be run unattended. That means defaults, validation, global effects, and output formats must be explicitly managed. You're 90% there — a few refinements will make this bulletproof for production usage.

Sankara-Jefferson commented:
**File : cmd/run/unified_service.go**

 ***Must Be Changed (Before Merge)***

**Workflow Monitor and Status Logger Use Fixed Tickers**
`runWorkflowMonitor() and printServiceStatus()` both use hardcoded intervals (30s, 2m) for critical logic.
Fix: Respect --check-interval for both functions (already parsed into orchestratorConfig).
 Why: A single global config flag should control all cadence-related behaviors to avoid confusion and misalignment.
**Missing Error Handling for Find(&jobCounts)**
In printServiceStatus(), the GORM Find() call does not check for errors.
Fix: Capture and log the error from .Find(&jobCounts).
Why: Without this, any SQL issue silently leads to incorrect or partial metrics in service logs.
**Orchestrator State Change Visibility**
Currently, there's no startup log that reflects which orchestration sub-stages (scan → pack, pack → daggen, etc.) are enabled or disabled.
Fix: Log orchestration config (at least the key booleans) at startup.
Why: Improves observability and gives immediate operator feedback.

***Nice To Have (Non-blocking)***

**Graceful Recovery on Panic in Goroutines**
The workflow and status ticker goroutines don’t recover from unexpected panics.
Suggestion: Wrap those goroutines with a defer func(){ recover() }() to avoid crashing the unified service loop.
**Dynamic Reload of Confi**g
Future consideration: add a signal handler or periodic config refresh to allow live tuning of scaling parameters.
Benefit: Enables long-lived services to adapt without restart.
**Expose Health Endpoint or Liveness Prob**e
If this service will run long-term in an orchestrator (K8s, Nomad), consider exposing a simple HTTP healthcheck in a future revision.
Benefit: Allows integration with readiness/liveness probes for auto-healing.
**Log Service Start Summary**
Currently it logs “Starting unified auto-preparation service” but doesn’t summarize key operational parameters.
Benefit: Single-line overview of worker limits, orchestration state, and scan/pack/daggen flags helps confirm expected startup behavior.

This is strong infra glue — it connects lifecycle, orchestration, and scaling well. What it needs now is tighter observability and shared configuration consistency so operators can trust what’s running without guesswork.

Sankara-Jefferson commented:
**File : handler/dataprep/autodeal.go**

***Must Be Changed (Before Merge)***
**Missing Transaction Wrapping**
Methods like CreateAutomaticDealSchedule and ProcessReadyPreparations perform multiple DB writes but lack transactional boundaries.
Fix: Wrap related operations in a db.Transaction() block (especially preparation fetch → schedule creation).
Why: Prevents partial state updates in case of mid-process failure (e.g., logging succeeds but schedule write fails).
**Silent Fallback on Provider Absence**
If DealProvider is unset and GetDefaultStorageProvider() fails, it logs but continues with deal creation.
Fix: Return early with a hard error if both are missing.
Why: Prevents silent attempts to create a schedule without a provider, which is likely to fail downstream.
**Validation Logic Breaks on First Wallet or SP Failure**
The validateWalletsForDealCreation() and validateProviderForDealCreation() functions return immediately on first failure.
Fix: Continue validating remaining items (collect all validation failures) before failing.
Why: Aggregated validation errors give more actionable feedback to users/admins.
**Logger Missing in ProcessReadyPreparations() Loop**
This method uses autoDealLogger.Debugf(...) directly but not the structured logInfo() or logError() methods for success/failure at the loop level.
Fix: Add a final logInfo call summarizing skipped vs. failed vs. created schedules.
Why: Keeps log streams consistent and auditable for automation observability.

***Nice To Have (Non-blocking)***

**Structured Logging Enhancements**
Log metadata (e.g., schedule ID, prep ID, provider) is embedded as model.ConfigMap — consider enriching with severity levels or timestamps for cross-service log correlation.
Benefit: Enables easier search/filtering if notifications/logs are ingested into a platform like Loki or Datadog.
**Testability Hooks**
You could expose AutoDealService.Config as a struct for future test overrides (e.g., mock validators, dry-run mode).
Benefit: Makes it easier to write integration tests or unit tests around auto-deal pipelines.

**Parallelization (Future)**
For large-scale prep sets, consider processing preparations in parallel (e.g., using errgroup.Group).
Benefit: Improves throughput for massive onboarding events.
**Return Detailed Error from ProcessReadyPreparations()**
Right now it returns a single error, but internally tracks errorCount. You could return a detailed summary error (e.g., struct with failed prep names).
Benefit: Improves UX in tools that call this method and want to present diagnostics.

This is strong backend logic for orchestrated automation — clearly designed with scale and composability in mind. What it needs now is transactional safety, better error aggregation, and defensive hard exits on missing deal-critical data. Logging and test scaffolding are already in a great place.

Sankara-Jefferson commented:
**File : handler/dataprep/create.go**

***Must Be Changed (Before Merge)***

Template Application Logic Ignores Non-Zero Request Overrides
Current logic in ApplyTemplateToPreparation() blindly overwrites values on preparation.DealConfig, even if the user specified non-default values in the request.
Fix: Only apply template defaults if corresponding request fields are unset/zero.
Why: Prevents unintentionally overriding explicit user intent (e.g., price, provider).
**Preparation ID Used Before DB Persistence**
In performValidation(), you're using preparation.ID in logs and validation metadata before the preparation has been .Create()'d in the DB.
Fix: Perform validation after the DB Create() call or generate a safe temporary ID string (e.g., "not-yet-persisted").
Why: Avoids misleading logs/notifications with missing or invalid preparation IDs.
**SP and Wallet Validation Currently Stubbed**
`performWalletValidation()` and `performSPValidation()` emit logs but skip actual Lotus validation logic.
Fix: Add TODO comments explicitly stating "Validation logic stubbed – full implementation pending integration with Lotus client".
Why: Makes it clear to reviewers and maintainers that this is an interim implementation.
**Silent Failure on Validation Errors**
Auto-deal creation proceeds even if wallet/SP validation fails — this is currently just logged as a warning.
Fix: Add a flag (failOnValidationErrors: true/false) to control behavior, or explicitly disable AutoCreateDeals in memory if validation fails.
Why: Prevents unintended automated deal creation due to silent validation skips.

***Nice To Have (Optional Polish)***

Template Application Logging: Log the template name and which values were applied from the template.
Testability Hint: Consider injecting notificationHandler, dealtemplateHandler, and spValidator for easier unit testing.
Field Reuse: Some validations use repeated string joins and field flattening — a reusable mapToConfigMap() helper could reduce repetition.
Consistency: Use strconv.FormatBool(...) instead of inline if/else for "true"/"false" string values.


Strong auto-deal integration into the prep handler, with well-structured validation and clear extension points. You’ll want to harden validation flow, preserve user intent during template application, and signal stubbed logic more clearly. Great evolution toward self-driving onboarding.

Sankara-Jefferson commented:
**File : handler/job/pack.go**

***Must Change Before Merge***

**Background goroutine error logging is too silent**
Why: Logging only a Warnf on auto-deal failure buries potential system misbehavior (e.g., missing schedules). This is a backend trigger — not a soft UX element.
Fix: At minimum, escalate this to an error log or emit a structured event via a notificationHandler like elsewhere in the code.

***Nice to Have (Not blocking)***

**Add metrics/logging for successful auto-deal triggers**
Why: Currently, there is no visibility into successful auto-deal triggers. This makes it difficult to audit whether the logic ever executes successfully in prod.
Improvement: Emit a success log or increment a Prometheus metric when the auto-deal trigger completes without error.
**Parameterize or document the 30s timeout**
Why: The goroutine uses a hardcoded 30 * time.Second timeout, which may or may not be suitable for all network conditions.
Improvement: Move to a constant, config value, or at least document the rationale for the chosen value.
**Consider error grouping for parallel job failures**
Why: If this logic is eventually used across multiple job types (e.g., DAGGen), consistent error handling/reporting for failed post-processing is important.
Improvement: Centralize autodeal errors into a structured handler.

anjor commented:
The latest feedback has been addressed.
